File: packages/js-sdk/tests/volume/mockVolumeContent.ts
import { http, HttpResponse } from 'msw'
import { randomUUID } from 'node:crypto'

import { apiUrl } from '../setup'

/**
 * In-memory mock of the volume control-plane CRUD plus the volume content API
 * (spec/openapi-volumecontent.yml), backing the `volumeTest` file-operation
 * tests with a stateful in-process filesystem.
 */

const DEFAULT_FILE_MODE = 0o644
const DEFAULT_DIR_MODE = 0o755

interface Entry {
  name: string
  type: 'file' | 'directory'
  path: string
  uid: number
  gid: number
  mode: number
  content: Uint8Array
  timestamp: string
}

// Volumes created via the control-plane mock, keyed by volume ID.
const volumes = new Map<
  string,
  { volumeID: string; name: string; token: string }
>()

// One in-memory filesystem per volume, keyed by volume ID.
const filesystems = new Map<string, Map<string, Entry>>()

function normalize(path: string): string {
  let normalized = path.trim()
  if (!normalized.startsWith('/')) {
    normalized = `/${normalized}`
  }
  // Collapse repeated slashes and strip a trailing slash (except for root).
  normalized = normalized.replace(/\/+/g, '/')
  if (normalized.length > 1 && normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1)
  }
  return normalized
}

function dirname(path: string): string {
  const idx = path.lastIndexOf('/')
  return idx <= 0 ? '/' : path.slice(0, idx)
}

function basename(path: string): string {
  return path.slice(path.lastIndexOf('/') + 1)
}

function makeEntry(
  path: string,
  type: Entry['type'],
  params: URLSearchParams,
  content: Uint8Array = new Uint8Array()
): Entry {
  const mode = params.get('mode')
  return {
    name: basename(path),
    type,
    path,
    uid: Number(params.get('uid') ?? 0),
    gid: Number(params.get('gid') ?? 0),
    mode:
      mode !== null
        ? Number(mode)
        : type === 'file'
          ? DEFAULT_FILE_MODE
          : DEFAULT_DIR_MODE,
    content,
    timestamp: new Date().toISOString(),
  }
}

function stat(entry: Entry) {
  return {
    name: entry.name,
    type: entry.type,
    path: entry.path,
    size: entry.content.length,
    mode: entry.mode,
    uid: entry.uid,
    gid: entry.gid,
    atime: entry.timestamp,
    mtime: entry.timestamp,
    ctime: entry.timestamp,
  }
}

function error(status: number, message: string) {
  return HttpResponse.json({ code: String(status), message }, { status })
}

function getFs(volumeID: string): Map<string, Entry> {
  let fs = filesystems.get(volumeID)
  if (!fs) {
    fs = new Map([['/', makeEntry('/', 'directory', new URLSearchParams())]])
    filesystems.set(volumeID, fs)
  }
  return fs
}

function makeDir(
  fs: Map<string, Entry>,
  path: string,
  params: URLSearchParams
) {
  const force = params.get('force') === 'true'
  const existing = fs.get(path)
  if (existing) {
    if (!force || existing.type !== 'directory') {
      return error(409, `Path ${path} already exists`)
    }
    return HttpResponse.json(stat(existing), { status: 201 })
  }

  const parentPath = dirname(path)
  const parent = fs.get(parentPath)
  if (!parent) {
    if (!force) {
      return error(404, `Path ${parentPath} not found`)
    }
    makeDir(fs, parentPath, params)
  } else if (parent.type !== 'directory') {
    return error(409, `Path ${parentPath} is not a directory`)
  }

  const entry = makeEntry(path, 'directory', params)
  fs.set(path, entry)
  return HttpResponse.json(stat(entry), { status: 201 })
}

export const mockVolumeApiHandlers = [
  // Control-plane volume CRUD used by the `volumeTest` fixture.
  http.post(apiUrl('/volumes'), async ({ request }) => {
    const { name } = (await request.clone().json()) as { name: string }
    const volumeID = randomUUID()
    const token = `vol-token-${randomUUID()}`
    volumes.set(volumeID, { volumeID, name, token })
    return HttpResponse.json({ volumeID, name, token }, { status: 201 })
  }),

  http.delete<{ volumeID: string }>(
    apiUrl('/volumes/:volumeID'),
    ({ params }) => {
      const existed = volumes.delete(params.volumeID)
      filesystems.delete(params.volumeID)
      if (!existed) {
        return error(404, 'Not found')
      }
      return new HttpResponse(null, { status: 204 })
    }
  ),

  // Volume content API.
  http.put<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/file'),
    async ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const query = new URL(request.url).searchParams
      const path = normalize(query.get('path') ?? '')

      const existing = fs.get(path)
      if (existing) {
        if (existing.type !== 'file') {
          return error(409, `Path ${path} is a directory`)
        }
        if (query.get('force') !== 'true') {
          return error(409, `Path ${path} already exists`)
        }
      }

      const parent = fs.get(dirname(path))
      if (!parent || parent.type !== 'directory') {
        return error(404, `Path ${path} not found`)
      }

      const content = new Uint8Array(await request.arrayBuffer())
      const entry = makeEntry(path, 'file', query, content)
      fs.set(path, entry)
      return HttpResponse.json(stat(entry), { status: 201 })
    }
  ),

  http.get<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/file'),
    ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const path = normalize(
        new URL(request.url).searchParams.get('path') ?? ''
      )
      const entry = fs.get(path)
      if (!entry || entry.type !== 'file') {
        return error(404, `Path ${path} not found`)
      }
      return new HttpResponse(
        entry.content.length > 0 ? entry.content.slice().buffer : null,
        {
          status: 200,
          headers: {
            'Content-Type': 'application/octet-stream',
            'Content-Length': String(entry.content.length),
          },
        }
      )
    }
  ),

  http.get<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/dir'),
    ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const path = normalize(
        new URL(request.url).searchParams.get('path') ?? ''
      )
      const entry = fs.get(path)
      if (!entry || entry.type !== 'directory') {
        return error(404, `Path ${path} not found`)
      }
      const prefix = path === '/' ? '/' : `${path}/`
      const children = Array.from(fs.entries())
        .filter(
          ([p]) =>
            p !== path &&
            p.startsWith(prefix) &&
            !p.slice(prefix.length).includes('/')
        )
        .map(([, e]) => stat(e))
      return HttpResponse.json(children)
    }
  ),

  http.post<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/dir'),
    ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const query = new URL(request.url).searchParams
      const path = normalize(query.get('path') ?? '')
      return makeDir(fs, path, query)
    }
  ),

  http.get<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/path'),
    ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const path = normalize(
        new URL(request.url).searchParams.get('path') ?? ''
      )
      const entry = fs.get(path)
      if (!entry) {
        return error(404, `Path ${path} not found`)
      }
      return HttpResponse.json(stat(entry))
    }
  ),

  http.patch<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/path'),
    async ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const path = normalize(
        new URL(request.url).searchParams.get('path') ?? ''
      )
      const entry = fs.get(path)
      if (!entry) {
        return error(404, `Path ${path} not found`)
      }
      const body = (await request.clone().json()) as {
        uid?: number
        gid?: number
        mode?: number
      }
      if (body.uid !== undefined) entry.uid = body.uid
      if (body.gid !== undefined) entry.gid = body.gid
      if (body.mode !== undefined) entry.mode = body.mode
      return HttpResponse.json(stat(entry))
    }
  ),

  http.delete<{ volumeID: string }>(
    apiUrl('/volumecontent/:volumeID/path'),
    ({ params, request }) => {
      const fs = getFs(params.volumeID)
      const path = normalize(
        new URL(request.url).searchParams.get('path') ?? ''
      )
      if (path === '/' || !fs.has(path)) {
        return error(404, `Path ${path} not found`)
      }
      const prefix = `${path}/`
      for (const p of Array.from(fs.keys())) {
        if (p === path || p.startsWith(prefix)) {
          fs.delete(p)
        }
      }
      return new HttpResponse(null, { status: 204 })
    }
  ),
]

export function resetMockVolumeApi() {
  volumes.clear()
  filesystems.clear()
}

File: packages/python-sdk/tests/mock_volume_content.py
"""In-memory mock of the volume content API (spec/openapi-volumecontent.yml).

Backs the `volume`/`async_volume` fixtures so volume file-operation tests run
against a stateful in-process filesystem instead of a live volume server.
"""

import json
import posixpath
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, Optional

import httpx

from e2b.volume.types import VolumeFileType

DEFAULT_FILE_MODE = 0o644
DEFAULT_DIR_MODE = 0o755


@dataclass
class _Entry:
    name: str
    type: VolumeFileType
    path: str
    uid: int = 0
    gid: int = 0
    mode: int = 0
    content: bytes = b""
    timestamp: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )

    def stat(self) -> dict:
        return {
            "name": self.name,
            "type": self.type,
            "path": self.path,
            "size": len(self.content),
            "mode": self.mode,
            "uid": self.uid,
            "gid": self.gid,
            "atime": self.timestamp,
            "mtime": self.timestamp,
            "ctime": self.timestamp,
        }


def _normalize(path: str) -> str:
    path = path.strip()
    if not path.startswith("/"):
        path = "/" + path
    return posixpath.normpath(path)


def _error(status: int, message: str) -> httpx.Response:
    return httpx.Response(
        status,
        content=json.dumps({"code": str(status), "message": message}).encode(),
        headers={"Content-Type": "application/json"},
    )


class MockVolumeContentAPI:
    """A per-test, in-memory volume filesystem behind an httpx.MockTransport."""

    def __init__(self):
        self.entries: Dict[str, _Entry] = {
            "/": _Entry(
                name="/",
                type=VolumeFileType.DIRECTORY,
                path="/",
                mode=DEFAULT_DIR_MODE,
            )
        }

    def handler(self, request: httpx.Request) -> httpx.Response:
        request.read()
        return self._handle(request)

    async def async_handler(self, request: httpx.Request) -> httpx.Response:
        await request.aread()
        return self._handle(request)

    def _handle(self, request: httpx.Request) -> httpx.Response:
        segments = request.url.path.strip("/").split("/")
        if len(segments) != 3 or segments[0] != "volumecontent":
            return _error(404, "Not found")
        resource = segments[2]
        params = request.url.params
        path = _normalize(params.get("path") or "")

        if resource == "file":
            if request.method == "GET":
                return self._read_file(path)
            if request.method == "PUT":
                return self._write_file(path, params, request.content)
        elif resource == "dir":
            if request.method == "GET":
                return self._list_dir(path)
            if request.method == "POST":
                return self._make_dir(path, params)
        elif resource == "path":
            if request.method == "GET":
                return self._stat(path)
            if request.method == "PATCH":
                return self._update_metadata(path, request.content)
            if request.method == "DELETE":
                return self._remove(path)

        return _error(404, "Not found")

    @staticmethod
    def _int_param(params: httpx.QueryParams, name: str) -> Optional[int]:
        value = params.get(name)
        return int(value) if value is not None else None

    @classmethod
    def _mode_param(cls, params: httpx.QueryParams, default: int) -> int:
        mode = cls._int_param(params, "mode")
        return mode if mode is not None else default

    @staticmethod
    def _bool_param(params: httpx.QueryParams, name: str) -> bool:
        return (params.get(name) or "").lower() == "true"

    def _read_file(self, path: str) -> httpx.Response:
        entry = self.entries.get(path)
        if entry is None or entry.type != VolumeFileType.FILE:
            return _error(404, f"Path {path} not found")
        return httpx.Response(
            200,
            content=entry.content,
            headers={"Content-Type": "application/octet-stream"},
        )

    def _write_file(
        self, path: str, params: httpx.QueryParams, content: bytes
    ) -> httpx.Response:
        existing = self.entries.get(path)
        if existing is not None:
            if existing.type != VolumeFileType.FILE:
                return _error(409, f"Path {path} is a directory")
            if not self._bool_param(params, "force"):
                return _error(409, f"Path {path} already exists")

        parent = self.entries.get(posixpath.dirname(path))
        if parent is None or parent.type != VolumeFileType.DIRECTORY:
            return _error(404, f"Path {path} not found")

        entry = _Entry(
            name=posixpath.basename(path),
            type=VolumeFileType.FILE,
            path=path,
            uid=self._int_param(params, "uid") or 0,
            gid=self._int_param(params, "gid") or 0,
            mode=self._mode_param(params, DEFAULT_FILE_MODE),
            content=content,
        )
        self.entries[path] = entry
        return httpx.Response(201, json=entry.stat())

    def _make_dir(self, path: str, params: httpx.QueryParams) -> httpx.Response:
        force = self._bool_param(params, "force")
        existing = self.entries.get(path)
        if existing is not None:
            if not force or existing.type != VolumeFileType.DIRECTORY:
                return _error(409, f"Path {path} already exists")
            return httpx.Response(201, json=existing.stat())

        parent_path = posixpath.dirname(path)
        parent = self.entries.get(parent_path)
        if parent is None:
            if not force:
                return _error(404, f"Path {parent_path} not found")
            res = self._make_dir(parent_path, params)
            if res.status_code >= 400:
                return res
        elif parent.type != VolumeFileType.DIRECTORY:
            return _error(409, f"Path {parent_path} is not a directory")

        entry = _Entry(
            name=posixpath.basename(path),
            type=VolumeFileType.DIRECTORY,
            path=path,
            uid=self._int_param(params, "uid") or 0,
            gid=self._int_param(params, "gid") or 0,
            mode=self._mode_param(params, DEFAULT_DIR_MODE),
        )
        self.entries[path] = entry
        return httpx.Response(201, json=entry.stat())

    def _list_dir(self, path: str) -> httpx.Response:
        entry = self.entries.get(path)
        if entry is None or entry.type != VolumeFileType.DIRECTORY:
            return _error(404, f"Path {path} not found")
        prefix = "/" if path == "/" else path + "/"
        children = [
            e.stat()
            for p, e in self.entries.items()
            if p != path and p.startswith(prefix) and "/" not in p[len(prefix) :]
        ]
        return httpx.Response(200, json=children)

    def _stat(self, path: str) -> httpx.Response:
        entry = self.entries.get(path)
        if entry is None:
            return _error(404, f"Path {path} not found")
        return httpx.Response(200, json=entry.stat())

    def _update_metadata(self, path: str, content: bytes) -> httpx.Response:
        entry = self.entries.get(path)
        if entry is None:
            return _error(404, f"Path {path} not found")
        body = json.loads(content or b"{}")
        if body.get("uid") is not None:
            entry.uid = body["uid"]
        if body.get("gid") is not None:
            entry.gid = body["gid"]
        if body.get("mode") is not None:
            entry.mode = body["mode"]
        return httpx.Response(200, json=entry.stat())

    def _remove(self, path: str) -> httpx.Response:
        if path == "/" or path not in self.entries:
            return _error(404, f"Path {path} not found")
        prefix = path + "/"
        for p in list(self.entries):
            if p == path or p.startswith(prefix):
                del self.entries[p]
        return httpx.Response(204)
